README.md
# square-limit

My take on the MC Escher square limit. Built wiht haskell `diagrams`.

Resources:
- "Functional Geometry" Peter Henderson, 2002
- https://einarwh.wordpress.com/2017/07/22/picture-combinators-and-recursive-fish/
- http://shashi.biz/ijulia-notebooks/funcgeo/
- another python example https://mapio.github.io/programming-with-escher/

ChangeLog.md
# Changelog for square-limit

## Unreleased changes


src/SVGLoader.hs
module SVGLoader where

import qualified Graphics.Svg as Svg
import System.IO.Unsafe

import Diagrams.Prelude
import Diagrams.Backend.SVG.CmdLine

import qualified Graphics.Svg.Types as Svg.Types

svgFish = unsafePerformIO $ maybe (error "no fish") id <$> (Svg.loadSvgFile "fish.svg")

emptyDiag = (lw none $ square 16)

fishToDiagram :: Svg.Document -> Diagram B
fishToDiagram doc = foldl' (\r a -> r <> drawElement a) emptyDiag (Svg.Types._elements doc)

drawElement :: Svg.Tree -> Diagram B 
drawElement (Svg.Types.PathTree (Svg.Path attrs definitions)) = drawSVGPath definitions
drawElement (Svg.Types.GroupTree (Svg.Group attrs children viewBox aspectRatio)) = 
             foldl' (\r a -> r <> drawElement a) emptyDiag children
drawElement e = error $ "can't handle " <> (show e) <> "(yet)"

-- renderSVG "meout.svg" (dims (r2 (800.0, 800.0))) (fishToDiagram svgFish)
-- reverse engineer Diagram => SVG => Diagram?
-- need to fold over the path
drawSVGPath :: [Svg.Types.PathCommand] -> Diagram B
drawSVGPath xs = strokePath $ resultingPath $ foldl' drawSVGPathCommand emptyPathContext xs


data PathContext n = PathContext { 
                      current :: (Point V2 n) 
                    , resultingPath :: (Path V2 n)
                    }
emptyPathContext = PathContext (P $ r2 (0, 0)) mempty

drawSVGPathCommand :: (OrderedField n) => PathContext n -> Svg.Types.PathCommand -> PathContext n
drawSVGPathCommand = undefined

{-

MoveTo !Origin ![RPoint]	

M or m command
LineTo !Origin ![RPoint]	

Line to, L or l Svg path command.
HorizontalTo !Origin ![Coord]	
Equivalent to the H or h svg path command.

VerticalTo !Origin ![Coord]	
Equivalent to the V or v svg path command.

CurveTo !Origin ![(RPoint, RPoint, RPoint)]	
Cubic bezier, C or c command

SmoothCurveTo !Origin ![(RPoint, RPoint)]	
Smooth cubic bezier, equivalent to S or s command

QuadraticBezier !Origin ![(RPoint, RPoint)]	
Quadratic bezier, Q or q command

SmoothQuadraticBezierCurveTo !Origin ![RPoint]	
Quadratic bezier, T or t command

EllipticalArc !Origin ![(Coord, Coord, Coord, Bool, Bool, RPoint)]	
Eliptical arc, A or a command.

EndPath	 -- Close the path, Z or z svg path command.

-}


src/LimitExample.hs
module LimitExample where

import Prelude hiding (cycle)
import Diagrams.Prelude
import Diagrams.Backend.SVG.CmdLine -- ! hmm

-- * from diagrams example package 
someFunc :: IO ()
someFunc = putStrLn "someFunc"

blank :: Diagram B
blank = lw none $ square 16

makeTile :: [[P2 Double]] -> Diagram B
makeTile = showOrigin . lw thin . centerXY . mconcat . map fromVertices


data Picture
  = Blank
  | Single [[P2 Double]]
  | Rot Picture
  | Cycle Picture
  | HPair Picture Picture
  | VPair Picture Picture
  | Quartet Picture Picture Picture Picture
  | SkewQuartet Picture Picture Picture Picture

----

rot p = rotateBy (1/4) p
cycle p = quartet (rot p) p (rot $ rot p) (rot $ rot $ rot p)
hpair p q = scaleX (1/2) $ centerXY (p ||| q)
vpair p q = scaleY (1/2) $ centerXY (p === q)
quartet p q r s = scale (1/2) $ centerXY ((p ||| q) === (r ||| s))
skewquartet p q r s = scale (1/3) $ centerXY ((scaleY 2 p ||| scale 2 q)
                                              === (r ||| scaleX 2 s))

-- fold over the Picture
drawPicture :: Picture -> Diagram B
drawPicture Blank = blank
drawPicture (Single m) = makeTile m
drawPicture (Rot p) = rot (drawPicture p)
drawPicture (Cycle p) = cycle (drawPicture p)
drawPicture (HPair p q) = hpair (drawPicture p) (drawPicture q)
drawPicture (VPair p q) = vpair (drawPicture p) (drawPicture q)
drawPicture (Quartet p q r s)
  = quartet (drawPicture p) (drawPicture q) (drawPicture r) (drawPicture s)
drawPicture (SkewQuartet p q r s)
  = skewquartet (drawPicture p) (drawPicture q) (drawPicture r) (drawPicture s)

-- prune
prune :: Double -> Picture -> Picture
prune s p = prune' (s,s) p where
  prune' (x,y) p | min x y < 1 = Blank -- cut off when either factor <1
  prune' (x,y) Blank = Blank
  prune' (x,y) (Single m) = Single m
  prune' (x,y) (Rot p) = Rot (prune' (y,x) p)
  prune' (x,y) (Cycle p) = Cycle (prune' (x/2,y/2) p) -- assumes x==y
  prune' (x,y) (HPair p q) = HPair (prune' (x/2,y) p) (prune' (x/2,y) q)
  prune' (x,y) (VPair p q) = VPair (prune' (x,y/2) p) (prune' (x,y/2) q)
  prune' (x,y) (Quartet p q r s)
    = Quartet (prune' x2y2 p) (prune' x2y2 q) (prune' x2y2 r) (prune' x2y2 s)
      where x2y2 = (x/2,y/2)
  prune' (x,y) (SkewQuartet p q r s)
    = SkewQuartet (prune' (x3,y32) p) (prune' (x32,y32) q)
                  (prune' (x3,y3) r) (prune' (x32,y3) s)
      where (x3,y3,x32,y32) = (x/3,y/3,x3*2,y3*2)

fishP = Single markingsP
fishQ = Single markingsQ
fishR = Single markingsR
fishS = Single markingsS

fishT = Quartet fishP fishQ fishR fishS
fishU = Cycle (Rot fishQ)

corner = SkewQuartet p q r s where
  p = VPair p' fishT
  p' = HPair (Rot s) p
  q = Quartet p' q fishU s'
  r = fishQ
  s = HPair fishT s'
  s' = VPair s (Rot (Rot (Rot p)))

squarelimit = Cycle corner

example = drawPicture (prune 24 squarelimit) `atop` square 14.5
         # fc darkseagreen
         # opacity 0.5
markingsP, markingsQ, markingsR, markingsS :: [[P2 Double]]
--- marking defines fish with nearby fish ...
markingsP = [
   [ (4^&4), (6^&0) ],
   [ (0^&3), (3^&4), (0^&8), (0^&3) ],
   [ (4^&5), (7^&6), (4^&10), (4^&5) ],
   [ (11^&0), (10^&4), (8^&8), (4^&13), (0^&16) ],
   [ (11^&0), (14^&2), (16^&2) ],
   [ (10^&4), (13^&5), (16^&4) ],
   [ (9^&6), (12^&7), (16^&6) ],
   [ (8^&8), (12^&9), (16^&8) ],
   [ (8^&12), (16^&10) ],
   [ (0^&16), (6^&15), (8^&16), (12^&12), (16^&12) ],
   [ (10^&16), (12^&14), (16^&13) ],
   [ (12^&16), (13^&15), (16^&14) ],
   [ (14^&16), (16^&15) ]
   ]

markingsQ = [
   [ (2^&0), (4^&5), (4^&7) ],
   [ (4^&0), (6^&5), (6^&7) ],
   [ (6^&0), (8^&5), (8^&8) ],
   [ (8^&0), (10^&6), (10^&9) ],
   [ (10^&0), (14^&11) ],
   [ (12^&0), (13^&4), (16^&8), (15^&10), (16^&16), (12^&10), (6^&7), (4^&7), (0^&8) ],
   [ (13^&0), (16^&6) ],
   [ (14^&0), (16^&4) ],
   [ (15^&0), (16^&2) ],
   [ (0^&10), (7^&11) ],
   [ (9^&12), (10^&10), (12^&12), (9^&12) ],
   [ (8^&15), (9^&13), (11^&15), (8^&15) ],
   [ (0^&12), (3^&13), (7^&15), (8^&16) ],
   [ (2^&16), (3^&13) ],
   [ (4^&16), (5^&14) ],
   [ (6^&16), (7^&15) ]
   ]

markingsR = [
   [ (0^&12), (1^&14) ],
   [ (0^&8), (2^&12) ],
   [ (0^&4), (5^&10) ],
   [ (0^&0), (8^&8) ],
   [ (1^&1), (4^&0) ],
   [ (2^&2), (8^&0) ],
   [ (3^&3), (8^&2), (12^&0) ],
   [ (5^&5), (12^&3), (16^&0) ],
   [ (0^&16), (2^&12), (8^&8), (14^&6), (16^&4) ],
   [ (6^&16), (11^&10), (16^&6) ],
   [ (11^&16), (12^&12), (16^&8) ],
   [ (12^&12), (16^&16) ],
   [ (13^&13), (16^&10) ],
   [ (14^&14), (16^&12) ],
   [ (15^&15), (16^&14) ]
   ]

markingsS = [
   [ (0^&0), (4^&2), (8^&2), (16^&0) ],
   [ (0^&4), (2^&1) ],
   [ (0^&6), (7^&4) ],
   [ (0^&8), (8^&6) ],
   [ (0^&10), (7^&8) ],
   [ (0^&12), (7^&10) ],
   [ (0^&14), (7^&13) ],
   [ (8^&16), (7^&13), (7^&8), (8^&6), (10^&4), (16^&0) ],
   [ (10^&16), (11^&10) ],
   [ (10^&6), (12^&4), (12^&7), (10^&6) ],
   [ (13^&7), (15^&5), (15^&8), (13^&7) ],
   [ (12^&16), (13^&13), (15^&9), (16^&8) ],
   [ (13^&13), (16^&14) ],
   [ (14^&11), (16^&12) ],
   [ (15^&9), (16^&10) ]
   ]

app/Main.hs
{-# LANGUAGE NoMonomorphismRestriction #-}
{-# LANGUAGE FlexibleContexts          #-}
{-# LANGUAGE TypeFamilies              #-}

module Main where

import Diagrams.Prelude
import Diagrams.Backend.SVG.CmdLine
import Diagrams.Backend.SVG(loadImageSVG)
import System.IO.Unsafe

import qualified LimitExample

myCircle :: Diagram B
myCircle = circle 1

letterF' = polygon ( with
   & polyType .~ PolySides
   [cw, ccw, ccw, cw, cw, ccw, ccw, ccw, ccw]
   [2,1,1,1, 1, 2,  1,   3, 5]
   )
   where
     ccw = 90 @@ deg
     cw  = -90 @@ deg

letterF = text "F" <> square 1 # lw none

almostClosed :: Trail' Line V2 Double
almostClosed = fromVertices $ (map p2
  [( 0.30, 0.20), 
  ( 0.40, 0.20),
  ( 0.40, 0.45),
  ( 0.60, 0.45),
  ( 0.60, 0.55),
  ( 0.40, 0.55),
  ( 0.40, 0.70),
  ( 0.70, 0.70),
  ( 0.70, 0.80),
  ( 0.30, 0.80)])

frotated1 = rotateAround  (p2 (0.5, 0.5)) (45 @@ deg) letterF

closedLine  = pad 1.1 . center  . hsep 1
  $ [ --almostClosed # strokeLine
     almostClosed # closeLine # wrapLoop # strokeTrail --strokeLoop
    ]

-- 'escher' functions

turn' = rotate (90 @@ deg)
flip' = reflectX
toss  =   ( translate (0.5 * unitY))  . scale (1.0 / sqrt 2) . rotate (45 @@ deg)
-- scale (1.0 / sqrt 2) . rotateAround( p2 (0.0, 0.5)) (45 @@ deg)
-- above (beside (turn (turn (flip p))) (turn (turn p))) (beside (flip p) p)
foo = ( (((turn' . turn' . flip') theFish') ||| ((turn' . turn') theFish')) # centerXY # showOrigin )
       ===
       (((flip' theFish') ||| theFish') # centerXY # showOrigin )

theFish = unsafePerformIO $ loadImageSVG "fish2.png"

-- FIXME: translate! moveOriginBy v === translate (negated v).
-- position relative to unit square 
-- (unitSquare  # fc orange ) <> 
theFish' = --(unitSquare # dashingN [0.02] 0 ) <> 
           (( moveOriginBy (0.125 * unitX) (scaleToY 1 $ scaleToX 1.5 theFish )) 
              # rectEnvelope (p2 (-0.75,-0.5)) (r2 (1.25, 1.0)) {- -})  # showOrigin
             -- # clipBy (rect 1.25 1.0 # moveOriginBy (0.125 * unitX)) # showOrigin)
clipToUnit = rectEnvelope (p2 (-0.5,-0.5)) (r2 (1.0, 1.0))
twoFishes =  (( turn' . turn') theFish') <> ( theFish')

ttile f =    (f <> (fishN <> fishE) )
  where
    fishN = f # toss # flip'
    fishE = fishN # turn' # turn' # turn' 

utile f =    ( (fishN <> fishW) <> (fishE <> fishS)) 
  where

   fishN = f # toss # flip'
   fishW = fishN # turn'
   fishS = fishW # turn'
   fishE = fishS # turn'
   
-- henderson => h e n 
--              d e r
--              s o n
-- scale 1/6?
nonet p q r s t u v w x = scale (1/3) $ position $ zip 
                          ( p2 <$> [(-1, 1), (0, 1), (1, 1),
                                    (-1, 0), (0, 0), (1, 0),
                                    (-1, -1), (0, -1), (1, -1)
                                   ]) 
                          [p, q, r, s, t, u, v, w, x]
-- nonet p q r s t u v w x = ( p `beside'` q `beside'` r ) 
--                               `above'`
--                           ( s `beside'` t `beside'` u ) 
--                               `above'`
--                           ( v `beside'` w `beside'` x )       
-- scaleY 0.5 $  position $ zip (map p2 [(0, -0.5), (0, 0.5)]) [b, a]

--how to position to bounding box? (p ||| q)  === (r ||| s)
quartet p q r s = (p `beside'` q) `above'` (r `beside'` s) 
---centerXY $ position $ zip (map p2 [(0, 0.875), (1, 0.875), (0, 0), (1, 0)]) [p,q,r,s]

-- this is not the way, brother
--position $ zip (map p2 [(0.25, 0.75), (0.75, 0.75), (0.25, 0.25), (0.75, 0.25)]) [p,q,r,s]

side n p = quartet  s s (t # turn') t
  where
   s = if n == 1 then blank else side (n - 1) p
   t = ttile p
  
corner n p = quartet c s (s # turn') u
  where 
    (c,s) = if n == 1 then (blank, blank) 
             else (corner (n - 1) p, side (n - 1) p)
    u = utile p
  

blank = mempty

limit n fish = nonet  cornerNW sideN cornerNE 
                      sideW center sideE 
                      cornerSW sideS cornerSE
  where
    cornerNW = corner n fish
    cornerSW = turn' cornerNW
    cornerSE = turn' cornerSW
    cornerNE = turn' cornerSE
    sideN = side n fish  
    sideW = turn' sideN
    sideS = turn' sideW
    sideE = turn' sideS
    center = utile fish                           
-- is our atop or <>, beside, |||, ===
-- # showOrigin
-- stack run  -- --selection circle -w 100 -h 100 -o output.svg && firefox output.svg

above' a b = scaleY 0.5 $  position $ zip (map p2 [(0, -0.5), (0, 0.5)]) [b, a]
beside' a b = (scaleX 0.5 $  position $ zip (map p2 [(-0.5, 0), (0.5, 0)]) [a, b])
                   
nonetTest = unitSquare <>  (nonet
       (circle 1) (circle 1) (circle 1)
       (circle 1) (circle 1) (circle 1)
       (circle 1) (circle 1) (circle 1))
  where
    letter x = unitSquare <> (text x #scaleToX  (1.0) #scaleToY (1.0) #showOrigin )

-- wow, it's working! and I'm not an idiot. 
-- the only problem is that diagrams can't read svg files, it's not exactly my problem,
-- but how hard is to add svg support if most of the primitives already working, 
-- and actually are compiled into an svg by default??
-- TODO: outline !!

triangle' = (fromOffsets $ (map r2
  [(0.5, -0.5), (-1, 0), (0, 1), (0.5, -0.5)])) # strokeLine 

main :: IO ()
main = multiMain [
          ("circle", myCircle)
        , ("letterF", letterF)
        , ("closedLine", closedLine)
        , ("frotated1", foo)
        , ("twoFishes", twoFishes)
        , ("toss", unitSquare <> toss (unitSquare <> letterF))
        , ("justFish", theFish')
        , ("justTriangle",  (triangle') # showOrigin )
        , ("ttile", ttile theFish')
        , ("utile", square 1 <> utile theFish')
        , ("nonet", nonetTest)
        , ("side", side 2 (theFish'))
        , ("corner", corner 2 theFish')
        , ("above", square 1 <> above' theFish' theFish')
        , ("limit", limit 2 theFish')
        , ("example", LimitExample.example)
        , ("triangle", limit 3 $ triangle')
        ]


fishP = makeTile markingsP

makeTile :: [[P2 Double]] -> Diagram B
makeTile = showOrigin . lw thin . centerXY . mconcat . map fromVertices

markingsP = [
  [ (4^&4), (6^&0) ],
  [ (0^&3), (3^&4), (0^&8), (0^&3) ],
  [ (4^&5), (7^&6), (4^&10), (4^&5) ],
  [ (11^&0), (10^&4), (8^&8), (4^&13), (0^&16) ],
  [ (11^&0), (14^&2), (16^&2) ],
  [ (10^&4), (13^&5), (16^&4) ],
  [ (9^&6), (12^&7), (16^&6) ],
  [ (8^&8), (12^&9), (16^&8) ],
  [ (8^&12), (16^&10) ],
  [ (0^&16), (6^&15), (8^&16), (12^&12), (16^&12) ],
  [ (10^&16), (12^&14), (16^&13) ],
  [ (12^&16), (13^&15), (16^&14) ],
  [ (14^&16), (16^&15) ]
  ]